webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java

```diff
@@ -0,0 +1,190 @@
+     * @return
+     */
+    @Override
+    @Timed("MicrosoftMTEngine.translate")
+    public TranslationsResponse translate(
+            List<String> textSources,
+            String sourceLocaleTag,
+            List<String> targetLocaleTags,
+            String sourceMimeType,
+            String customModel) {
+
+        HttpHeaders headers = getHttpHeaders();
+
+        MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
+        params.put("api-version", Collections.singletonList(API_VERSION));
+        params.put("to", targetLocaleTags);
+
+        if (Strings.isNotBlank(sourceLocaleTag)) {
+            params.put("from", Collections.singletonList(sourceLocaleTag));
+        }
+
+        if (Strings.isNotBlank(sourceMimeType)) {
+            params.put("textType", Collections.singletonList(sourceMimeType));
+        }
+
+        if (Strings.isNotBlank(customModel)) {
+            params.put("domain", Collections.singletonList(customModel));
+        }
+
+        String finalUri =
+                UriComponentsBuilder.fromHttpUrl(getUrl(API_TRANSLATE))
+                        .queryParams(params)
+                        .build()
+                        .toUriString();
+
+        // TODO (garion) enforce API limit and implement batching
+
+        List<Map<String, String>> body =
+                textSources.stream()
+                        .map(text -> Collections.singletonMap("text", text))
+                        .collect(Collectors.toList());
+
+        HttpEntity<List<Map<String, String>>> requestEntity = new HttpEntity<>(body, headers);
+
+        try {
+            MicrosoftTextTranslationDTO[] response =
+                    restTemplate.postForObject(
+                            finalUri, requestEntity, MicrosoftTextTranslationDTO[].class);
+
+            return getTranslationsResponseDTO(textSources, response, targetLocaleTags);
+        } catch (HttpClientErrorException e) {
+            // Log only the request body, as the headers contain authentication information.
+            String errorMessage =
+                    String.format(
+                            "Could not Machine Translate posted to: %s with %s: %s and request body: %s",
+                            finalUri,
+                            X_CLIENT_TRACE_ID,
+                            requestEntity.getHeaders().get(X_CLIENT_TRACE_ID),
+                            body);
+            logger.error(errorMessage, e);
+            throw e;
+        }
+    }
+
+    private TranslationsResponse getTranslationsResponseDTO(
+            List<String> textSources, MicrosoftTextTranslationDTO[] response, List<String> targetLocaleTags) {
+        List<TextUnitTranslationGroup> textUnitTranslationGroups = new ArrayList<>();
+
+        // Translation objects in the output for each source string are returned in the same order as provided by the client.
+        for (int i = 0; i < textSources.size(); i++) {
```

Comment:
nit: can probably do soemthing like Streams.zip(textSources.stream(), response.stream(), (ts, mt) -> ... ).collect(toList())

Reply:
'zip(java.util.stream.Stream<A>, java.util.stream.Stream<B>, java.util.function.BiFunction<? super A,? super B,R>)' is marked unstable with @Beta 

- are we still Ok to use it given that warning?

Reply:
I'd probably fine. looks like the beta was added later, like this old doc doesn't have it: https://guava.dev/releases/23.0/api/docs/com/google/common/collect/Streams.html